src/utils/resolveUser.ts

```diff
@@ -1,40 +1,42 @@
-import { OAuth2Client } from "google-auth-library";
-import { TokenPayload } from "google-auth-library/build/src/auth/loginticket";
 import * as jwt from "jsonwebtoken";
-import { User, UserSessions } from "../generated/prisma-client";
-import { Context } from "prisma-client-lib/dist/types";
+import { Prisma, User, Admin } from "../generated/prisma-client";
 
-const client = new OAuth2Client(process.env.GOOGLE_CLIENT_ID);
-
-async function resolveUserUsingGoogle(ctx): Promise<User> {
-  try {
-    await client.verifyIdToken({
-      idToken: ctx.token,
-      audience: process.env.GOOGLE_CLIENT_ID,
-    });
-  } catch (e) {
-    return null;
-  }
-  const ticket = await client.verifyIdToken({
-    idToken: ctx.token,
-    audience: process.env.GOOGLE_CLIENT_ID,
-  });
-  const payload: TokenPayload = ticket.getPayload();
-  if (payload && payload.hd !== "connect.hku.hk") return null;
-  return ctx.prisma.user({ email: payload.email });
+interface SessionObject extends Object {
+  id: string;
+  createdAt?: string;
+  iat: number;
 }
 
-async function resolveUserUsingJWT(ctx: Context): Promise<User> | null {
+export const resolveUserUsingJWT = async (
+  prisma: Prisma,
+  token: string
+): Promise<any> => {
   try {
-    const userSession: UserSessions = jwt.verify(
-      ctx.token,
+    const sessionObject = jwt.verify(
+      token,
       process.env.PRISMA_SECRET
-    ) as UserSessions;
-    const sessionId = userSession.id;
-    return ctx.prisma.userSessions({ id: sessionId }).user();
+    ) as SessionObject;
```

Comment:
Please walk me through this. `generateToken()` it takes in either a `UserSession` or an `AdminUserSession` to sign a `token`. So when a `token` is verified and converted back, it can be either of them, but is casted as a `SessionObject` where the `id` is important, and used to identify if the token is from a `UserSession` or `AdminUserSession` by searching the database.

Reply:
I was thinking shouldn't it be more elegant to use just a singular `Session` datamodel where it has both a `user` and `admin` field (`@relation` with either a `User` or an `Admin`, both nullable). And then check which one is not `undefined`, so we could tell what kind of `Session` this is. I just find it funny why should 2 kinds of `Session`s exist when they serve the same purpose.

Reply:
The issue I think is that Prisma does not allow storing `User | Admin`. And I think that adding a `user` and `admin` in one table defeats the purpose of trying to unify `Session` anyway? The logic remains the same where you will check whether it is an admin or user? Am I missing something? Let me know your idea in detail.